builtin/logical/pki/ocsp.go

```diff
@@ -0,0 +1,397 @@
+		if err != nil {
+			return nil, err
+		}
+		return buf.Bytes(), nil
+	default:
+		return nil, fmt.Errorf("unsupported request method: %s", request.HTTPRequest.Method)
+	}
+}
+
+func logAndReturnInternalError(b *backend, err error) *logical.Response {
+	// Since OCSP might be a high traffic endpoint, we will log at debug level only
+	// any internal errors we do get. There is no way for us to return to the end-user
+	// errors, so we rely on the log statement to help in debugging possible
+	// issues in the field.
+	b.Logger().Debug("OCSP internal error", "error", err)
+	return OcspInternalErrorResponse
+}
+
+func getOcspStatus(sc *storageContext, request *logical.Request, ocspReq *ocsp.Request) (*ocspRespInfo, error) {
+	revEntryRaw, err := fetchCertBySerialBigInt(sc.Context, sc.Backend, request, revokedPath, ocspReq.SerialNumber)
+	if err != nil {
+		return nil, err
+	}
+
+	info := ocspRespInfo{
+		serialNumber: ocspReq.SerialNumber,
+		ocspStatus:   ocsp.Good,
+	}
+
+	if revEntryRaw != nil {
+		var revEntry revocationInfo
+		if err := revEntryRaw.DecodeJSON(&revEntry); err != nil {
+			return nil, err
+		}
+
+		info.ocspStatus = ocsp.Revoked
+		info.revocationTimeUTC = &revEntry.RevocationTimeUTC
+		info.issuerID = revEntry.CertificateIssuer // This might be empty if the CRL hasn't been rebuilt
+	}
+
+	return &info, nil
+}
+
+func lookupOcspIssuer(sc *storageContext, req *ocsp.Request, optRevokedIssuer issuerID) (*certutil.ParsedCertBundle, error) {
+	reqHash := req.HashAlgorithm
+	if !reqHash.Available() {
+		return nil, x509.ErrUnsupportedAlgorithm
+	}
+
+	// This will prime up issuerIds, with either the optRevokedIssuer value if set
+	// or if we are operating in legacy storage mode, the shim bundle id or finally
+	// a list of all our issuers in this mount.
+	issuerIds, err := lookupIssuerIds(sc, optRevokedIssuer)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, issuerId := range issuerIds {
+		parsedBundle, issuer, err := getOcspIssuerParsedBundle(sc, issuerId)
+		if err != nil {
+			// A bit touchy here as if we get an ErrUnknownIssuerId for an issuer id that we picked up
+			// from a revocation entry, we still return an ErrUnknownOcspIssuer as we can't validate
+			// the end-user actually meant this specific issuer's cert with serial X.
+			if errors.Is(err, ErrUnknownIssuerId) || errors.Is(err, ErrIssuerHasNoKey) {
+				// This skips either bad issuer ids, or root certs with no keys that we can't use.
+				continue
+			}
+			return nil, err
+		}
+
+		// Make sure the client and Vault are talking about the same issuer, otherwise
+		// we might have a case of a matching serial number for a different issuer which
+		// we should not respond back in the affirmative about.
+		matches, err := doesRequestMatchIssuer(parsedBundle, req)
+		if err != nil {
+			return nil, err
+		}
+
+		if matches {
+			if !issuer.Usage.HasUsage(OCSPSigningUsage) {
+				// We found the correct issuer, but it's not allowed to sign the
```

Comment:
Is it not possible for there to be more than one matching issuer here? (Same name, key, different expiry/certificate usages?)

Reply:
Hmm, this is a good point.